optuna_dashboard/ts/components/GraphTimeline.tsx

```diff
@@ -106,14 +101,38 @@ const plotTimeline = (trials: Trial[], mode: string) => {
       y: bars.map((b) => b.number),
       // @ts-ignore: To suppress ts(2322)
       base: starts.map((s) => s.toISOString()),
-      name: s,
+      name: name,
       text: bars.map((b) => makeHovertext(b)),
-      hovertemplate: "%{text}<extra>" + s + "</extra>",
+      hovertemplate: "%{text}<extra>" + name + "</extra>",
       orientation: "h",
-      marker: { color: cm[s] },
+      marker: { color: color },
       textposition: "none", // Avoid drawing hovertext in a bar.
     }
-    traces.push(trace)
+    return trace
+  }
+
+  const traces: Partial<plotly.PlotData>[] = []
+  for (const s of Object.keys(cm) as TrialState[]) {
+    const bars = trials.filter((t) => t.state === s)
+    if (bars.length === 0) {
+      continue
+    }
+    if (s == "Complete") {
+      const feasibleTrials = bars.filter((t) =>
+        t.constraints.every((c) => c <= 0)
+      )
+      const infeasibleTrials = bars.filter((t) =>
+        t.constraints.some((c) => c > 0)
+      )
+      if (feasibleTrials.length > 0) {
+        traces.push(makeTrace(feasibleTrials, "Complete", cm[s]))
+      }
+      if (infeasibleTrials.length > 0) {
+        traces.push(makeTrace(infeasibleTrials, "Infeasible", "#cccccc"))
+      }
+    } else {
+      traces.push(makeTrace(bars, s, cm[s]))
+    }
```

Comment:
This may cause a conflict with @nabenabe0928's comment, but how about the following to avoid the type-assertion (`a as T`? Type assertions may still leave a possibility of a runtime error, even if the code compiles successfully. This is due to the fact that type assertions do not reference any information at runtime and simply change the type of variables.

```suggestion
  for (const [state, color] of Object.entries(cm)) {
    const bars = trials.filter((t) => t.state === state)
    if (bars.length === 0) {
      continue
    }
    if (state == "Complete") {
      const feasibleTrials = bars.filter((t) =>
        t.constraints.every((c) => c <= 0)
      )
      const infeasibleTrials = bars.filter((t) =>
        t.constraints.some((c) => c > 0)
      )
      if (feasibleTrials.length > 0) {
        traces.push(makeTrace(feasibleTrials, "Complete", color))
      }
      if (infeasibleTrials.length > 0) {
        traces.push(makeTrace(infeasibleTrials, "Infeasible", "#cccccc"))
      }
    } else {
      traces.push(makeTrace(bars, state, color))
    }
  }
```